damdam6/일급객체와 javascript/일급객체와 javascript.md

```diff
@@ -0,0 +1,141 @@
+# 일급객체와 javascript
+
+### 일급 객체란?
+
+<aside>
+💡 A first class object is an entity that can be dynamically created, destroyed, passed to a function, returned as a value, and have all the rights as other variables in the programming language have.
+
+</aside>
+
+- 일급 객체의 조건
+    1. All items can be the actual parameters of functions
+        
+        함수의 매개변수가 될 수 있다.
+        
+    2. All items can be returned as results of functions
+        
+        함수의 결과로 반환될 수 있다.
+        
+    3. All items can be the subject of assignment statements
+        
+        할당문의 대상이 될 수 있다. (변수 등에 할당할 수  있다.)
+        
+    4. All items can be tested for equality
+        
+        서로 같은지 비교할 수 있다.
```

Comment:
4번 조건은 좀 생소한데, 예를 들어서 설명해줄 수 있나요?

Reply:
```
Robin Popplestone gave the following definition: All items have certain fundamental rights.

All items can be the actual parameters of functions
All items can be returned as results of functions
All items can be the subject of assignment statements
All items can be tested for equality.
```

```
Depending on the language, this can imply:

being expressible as an anonymous literal value
being storable in variables
being storable in data structures
having an intrinsic identity (independent of any given name)
being comparable for equality with other entities
being passable as a parameter to a procedure/function
being returnable as the result of a procedure/function
being constructible at runtime
being printable
being readable
being transmissible among distributed processes
being storable outside running processes
```

영문으로 된 위키와 stackoverflow에서 참고한 자료에 기반하고 있습니다.

* `All items can be tested for equality.`
* `being comparable for equality with other entities`

두 가지가 유사한 의미를 담고 있습니다. `다른 엔티티와 비교 가능하다`입니다. 이는 객체를 서로 동일한지 확인할 수 있음을 의미합니다. (ex. 1==1) 원시값들은 이를 충족시킬 수 있겠죠.

그러나 객체의 참조 비교를 하게 되는 경우에는동일성을 비교할 수 없습니다. 이러한 이유 때문에 이 부분은 일급객체의 조건보다는 **특징(언어마다 다를 수 있음)**에 포함하는 게 좋을 것 같다고 생각되긴 합니다... 

4번 조건 일단 삭제하도록 하겠습니다!